lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -4619,6 +4619,12 @@ void SROA_Parameter_HLSL::flattenGlobal(GlobalVariable *GV) {
 
     // Flat Global vector if no dynamic vector indexing.
     bool bFlatVector = !hasDynamicVectorIndexing(EltGV);
+
+    // Disable scalarization of groupshared vector arrays
```

Comment:
Ok, but why?

Reply:
Data locality of vector tgsm arrays is destroyed when splitting array of vector into one scalar array per vector component, leading to performance problems.  It is also difficult for a back end compiler to recover this information, once lost.